### ISSUE TITLE
REST/Serial over wifi

### ISSUE THREAD

**Issue by BlipRanger** [jkua/wheelwriter-interface#3]
This issue is just to check in on the current state of / discuss the REST api implementation. Is the plan to essentially run a REST server which can call the same commands exposed via serial?

**Comment by jkua** [comment]
Unfortunately REST support currently doesn't exist; it is merely a figment of my imagination. That said, it sounds like a good weekend project, so I might take a look this weekend.

I'd be interested to hear what you're looking for in such an interface. My thinking was that this would a convenient way for smartphones, tablets, and other devices on the network to talk to the typewriter.

My initial thought for implementation was to provide REST calls for the some of the various functions currently implemented:

1. `type` would be a good first candidate as you can just send it ASCII + escape sequences and it will print it
2. `relay` would certainly be powerful, but developers would need to port some/all of Wheelwriter.h in order to use it.
3. `read` and `keyboard` seems messy for REST. You'd have to poll to get data
4. `query`, `sample`, and the tests `buffer`, `char`, `circle` would be fine. The tests probably have limited utility

A lock would need to be implemented so that commands from multiple users, whether serial or REST, won't step on each other

**Comment by jkua** [comment]
I've implemented initial REST support! This supports the following POST endpoints:

1. `/type` - send an ASCII file to this endpoint and the typewriter will type it. Supports the same [ANSI/CSI-style escape codes](https://github.com/jkua/wheelwriter-interface/blob/rest-v1/src/arduino/wheelwriter_interface/Wheelwriter.cpp#L774) that the serial console does.
2. `/query` - responds with a JSON listing the Wheelwriter model, printwheel pitch, and status
3. `/characterTest`, `/circleTest`,  and `/printwheelSample`

If you install [rest-v1](https://github.com/jkua/wheelwriter-interface/tree/rest-v1), you can use the serial console to setup the WiFi with the `wifi` command. The IP address will be listed and you should get a simple web page if you access that in a browser.

* `curl -X POST http://<ip_address>/type -d "test1234"` should cause your Wheelwriter to type `test1234`
* `curl -X POST http://192.168.2.87/type -d "test^[[1m1234"` will output the same, but `1234` will be bold
* `curl -X POST http://<ip_address>/type --data-binary "@<filename>"` to send a text file
* `curl -X POST http://<ip_address>/query` should return something like `{"model":6,"wheel":32,"status":0}`

**Comment by jkua** [comment]
Hmm, I could implement a read function that waits for a carriage return. I can see the utility in that.

It would be blocking, just as all the other functions are, but I'll add a timeout parameter. It would only listen for keyboard entry during the read command, sorta half-duplex. Full duplex communication would probably require a man-in-the-middle setup and a running input buffer.

I'll see if I can implement that this morning!

**Comment by jkua** [comment]
OK, I added `/readLine` in [rest-v1.2](https://github.com/jkua/wheelwriter-interface/tree/rest-v1.2) - this reads a line until return is pressed. 

There is a configurable `timeout` (sent in the [POST request to the endpoint](https://github.com/jkua/wheelwriter-interface/blob/rest-v1.2/docs/wwib_rest_api.md)) that sets a timeout in integer milliseconds from the last key pressed, so that if someone partially enters a line and never hits return, you'll get it.

By default, you'll get the corrected line if the user backspaces. You can set the `corrected` parameter to `0` to get the uncorrected line with backspace characters included if you want to see what the user edited.

Let me know what you think!

**Comment by jkua** [comment]
Yeah, I've got a good idea why that's happening. I suspect it has to do with how I interpret reverse carriage movement commands as backspace and platen movement as line feed. I'm actually working on a direct way to get the key pressed, rather than this indirect way of looking at movement commands, but that's a larger change.

I can probably write a patch, but before I do, what version of the code are you using?
